### PR TITLE
[BW Clock] use widget_utils

### PR DESCRIPTION
**Pull request by thyttan**
#2478 #2196

Works well for me with Screen: Normal, Dynamic and Full.

### CODE REVIEW

File: apps/bwclk/app.js [hunk 1]
```diff
@@ -277,9 +277,9 @@ let drawLock = function() {
 
 let drawWidgets = function() {
   if(isFullscreen()){
-    for (let wd of WIDGETS) {wd.draw=()=>{};wd.area="";}
+    widget_utils.hide();
   } else {
-    Bangle.drawWidgets();
+    Bangle.drawWidgets(); // Should this be replaced with widget_utils.show(); ?
```
Comment (thyttan) [inline apps/bwclk/app.js:282]: Should this be replaced with widget_utils.show(); ?

File: apps/bwclk/app.js [hunk 2]
```diff
@@ -370,7 +370,8 @@ Bangle.loadWidgets();
 
 // Cache draw function for dynamic screen to hide / show widgets
 // Bangle.loadWidgets() could also be called later on but its much slower!
-for (let wd of WIDGETS) {wd._draw=wd.draw; wd._area=wd.area;}
+//for (let wd of WIDGETS) {wd._draw=wd.draw; wd._area=wd.area;} // Not sure how to replace this or if it should just be dropped?
+widget_utils.show(); // I replace with this to begin with, should I just delete it?
```
Comment (thyttan) [inline apps/bwclk/app.js:374]: // Not sure how to replace this or if it should just be dropped?

widget_utils.show(); // I replace with this to begin with, should I just delete it?

File: apps/bwclk/app.js (hunk 1 above)
Comment (thyttan) [reply apps/bwclk/app.js:282]: Tried this on another clock app, didn't work very well so I'll not change to .show() here.

**Comment by gfwilliams** [comment]
Hi - please can you add `widget_utils.show()` in the remove handler?

That's the real issue here. Basically even if you weren't using widget_utils, just showing the widgets again in the remove handler would have fixed it.

As-is, even with these changes I think that if the clock fast-loads to a launcher that doesn't use widget_utils, the widgets will still stay hidden?

**Comment by thyttan** [comment]
Sorry! Just replaced already existing implementation and didn't think to do that even though I should have remembered that's how it should be done.

**Comment by gfwilliams** [comment]
No worries -  thanks!

**Comment by thyttan** [comment]
There, .show() in remove handler, also removed 371-374 as per my review comments above.

>edit: It'd be nice to have some way of testing this, but I'm not too sure how. I think we'd have to spin up an emulator, and it all gets a bit tricky.

Would it be possible to search for if the string ".show()" is between the setUI's curly braces? That could help many cases I think?

**Comment by gfwilliams** [comment]
Great, thanks!

> Would it be possible to search for if the string ".show()" is between the setUI's curly braces? 

It would be with a bit of work, but then I guess we'd have to check for if widget_utils was used - plenty of faces *don't* hide widgets.

But that is only one issue - there's loaded libraries, memory usage, etc - like what's in http://www.espruino.com/Bangle.js+Fast+Load - I'm not sure it's reasonable to expect everyone to test all that all the time.

I think as we get more fast-loaded apps someone making a change that breaks one is going to get more and more common, and it could end up being quite a big issue. I did start something at https://github.com/espruino/BangleApps/blob/master/bin/runapptests.js and maybe when I get time I'll see if I can add a fast-loading check in there

**Pull request merged**